File: columbu/CATscala/src/main/java/com/columbu/p2p/udp/message/discover/PingMessage.java
package com.galileo.p2p.udp.message.discover;

import static com.galileo.p2p.udp.message.UdpMessageTypeEnum.DISCOVER_PING;

import com.google.protobuf.ByteString;
import lombok.extern.slf4j.Slf4j;
import com.galileo.p2p.udp.message.Message;
import com.galileo.core.overlay.discover.node.Node;
import com.galileo.utils.ByteArray;
import com.galileo.core.config.args.Args;
import com.galileo.protos.protos.Discover;
import com.galileo.protos.protos.Discover.Endpoint;

@Slf4j
public class PingMessage extends Message {

  private Discover.PingMessage pingMessage;

  public PingMessage(byte[] data) throws Exception{
    super(DISCOVER_PING, data);
    this.pingMessage = Discover.PingMessage.parseFrom(data);
  }

  public PingMessage(Node from, Node to) {
    super(DISCOVER_PING, null);
    Endpoint fromEndpoint = Endpoint.newBuilder()
        .setNodeId(ByteString.copyFrom(from.getId()))
        .setPort(from.getPort())
        .setAddress(ByteString.copyFrom(ByteArray.fromString(from.getHost())))
        .build();
    Endpoint toEndpoint = Endpoint.newBuilder()
        .setNodeId(ByteString.copyFrom(to.getId()))
        .setPort(to.getPort())
        .setAddress(ByteString.copyFrom(ByteArray.fromString(to.getHost())))
        .build();
    this.pingMessage = Discover.PingMessage.newBuilder()
        .setVersion(Args.getInstance().getNodeP2pVersion())
        .setFrom(fromEndpoint)
        .setTo(toEndpoint)
        .setTimestamp(System.currentTimeMillis())
        .build();
    this.data = this.pingMessage.toByteArray();
  }

  public int getVersion(){
    return this.pingMessage.getVersion();
  }

  public Node getTo() {
    Endpoint to = this.pingMessage.getTo();
    Node node = new Node(to.getNodeId().toByteArray(),
        ByteArray.toStr(to.getAddress().toByteArray()), to.getPort());
    return node;
  }

  @Override
  public Node getFrom() {
    return Message.getNode(pingMessage.getFrom());
  }

  @Override
  public String toString() {
    return "[pingMessage: " + pingMessage;
  }

}

File: columbu/CATscala/src/main/scala/com/columbu/p2p/udp/message/discover/PingMessage.scala
package com.galileo.p2p.udp.message.discover

import com.google.protobuf.ByteString
import com.galileo.p2p.udp.message.Message
import com.galileo.p2p.discover.node.Node
import com.galileo.utils.ByteArray
import com.galileo.core.config.args.Args
import com.galileo.core.enumeration.UdpMessageTypeEnum
import com.galileo.protos.protos.Discover
import com.galileo.protos.protos.Discover.Endpoint

class PingMessage extends Message{

  private var pingMessage: Discover.PingMessage = null

  def this(data: Array[Byte]) {
    this()
    this.`type` = UdpMessageTypeEnum.DISCOVER_PING
    this.data = data
    this.pingMessage = Discover.PingMessage.parseFrom(data)
  }

  def this(from: Node, to: Node) {
    this()
    this.`type` = UdpMessageTypeEnum.DISCOVER_PING
    val fromEndpoint = Endpoint.newBuilder.setNodeId(ByteString.copyFrom(from.getId)).setPort(from.getPort).setAddress(ByteString.copyFrom(ByteArray.fromString(from.getHost))).build
    val toEndpoint = Endpoint.newBuilder.setNodeId(ByteString.copyFrom(to.getId)).setPort(to.getPort).setAddress(ByteString.copyFrom(ByteArray.fromString(to.getHost))).build
    this.pingMessage = Discover.PingMessage.newBuilder.setVersion(Args.getInstance.getNodeP2pVersion).setFrom(fromEndpoint).setTo(toEndpoint).setTimestamp(System.currentTimeMillis).build
    this.data = this.pingMessage.toByteArray
  }

  def getVersion: Int = this.pingMessage.getVersion

  def getTo: Node = {
    val to = this.pingMessage.getTo
    new Node(to.getNodeId.toByteArray, ByteArray.toStr(to.getAddress.toByteArray), to.getPort)
  }

  override def getFrom: Node = {
    Message.getNode(pingMessage.getFrom)
  }

  override def toString: String = "[pingMessage: " + pingMessage
}
